packages/table/src/TableContext.tsx

```diff
@@ -0,0 +1,314 @@
+import React, { createContext, useContext, useMemo } from 'react';
+
+const Types = {
+  SelectableTable: 'SELECTABLE_TABLE',
+  ToggleHeaderCheckedState: 'TOGGLE_HEADER_CHECKED',
+  SetColumnInfo: 'SET_COLUMN_INFO',
+  SortTableData: 'SORT_TABLE_DATA',
+  ToggleIndividualChecked: 'TOGGLE_INDIVIDUAL_CHECKED',
+  ToggleIndividualDisabled: 'TOGGLE_INDIVIDUAL_DISABLED',
+  SetHasNestedRows: 'SET_HAS_NESTED_ROWS',
+  SetHasRowSpan: 'SET_HAS_ROW_SPAN',
+  RegisterRow: 'REGISTER_ROW',
+} as const;
+
+type Types = typeof Types[keyof typeof Types];
+
+export { Types };
+
+interface ActionPayload {
+  [Types.SelectableTable]: boolean;
+  [Types.ToggleHeaderCheckedState]: undefined;
+  [Types.ToggleIndividualChecked]: {
+    index: number;
+    checked: boolean;
+  };
+  [Types.ToggleIndividualDisabled]: {
+    index: number;
+    disabled: boolean;
+  };
+  [Types.RegisterRow]: {
+    index: number;
+    checked: boolean;
+    disabled: boolean;
+  };
+  [Types.SetColumnInfo]: {
+    dataType?: DataType;
+    index: number;
+  };
+  [Types.SetHasRowSpan]: boolean;
+  [Types.SetHasNestedRows]: boolean;
+  [Types.SortTableData]: {
+    columnId: number;
+    accessorValue: () => string;
+    data: Array<unknown>;
+  };
+}
+
+type ActionMap<A extends Record<string, any>> = {
+  [Key in keyof A]: A[Key] extends undefined
+    ? {
+        type: Key;
+      }
+    : {
+        type: Key;
+        payload: A[Key];
+      };
+};
+
+type Action = ActionMap<ActionPayload>[keyof ActionMap<ActionPayload>];
+
+type Dispatch = (action: Action) => void;
+
+interface Sort {
+  columnId?: number;
+  direction: 'asc' | 'desc';
+  accessorValue?: () => string;
+}
+
+const DataType = {
+  Number: 'Number',
+  Quantity: 'quantity',
+  Weight: 'weight',
+  ZipCode: 'zipCode',
+  String: 'string',
+  Date: 'date',
+} as const;
+
+type DataType = typeof DataType[keyof typeof DataType];
+
+export { DataType };
+
+export interface State {
+  sort?: Sort;
+  data: Array<any>;
+  columnInfo?: {
+    [k in number]: { dataType?: DataType };
+  };
```

Comment:
I like using the shorthand `Record<number, { dataType?: DataType }>`

Reply:
👍